### ISSUE TITLE
Sik-Ho Tang | Review: Shake-Shake Regularization (Image Classification).

### ISSUE THREAD

**Issue by NorbertZheng** [NorbertZheng/read-papers#115]
Sik-Ho Tang. [Review: Shake-Shake Regularization (Image Classification)](https://towardsdatascience.com/review-shake-shake-regularization-image-classification-d22bb8587953).

**Comment by NorbertZheng** [comment]
## Overview
In this story, Shake-Shake Regularization (Shake-Shake), by Xavier Gastaldi from London Business School, is briefly reviewed. The motivation of this paper is that data augmentation is applied at the input image,
 - data augmentation might be also possible to apply data augmentation techniques to internal representations.

It is found in prior art that adding noise to the gradient during training helps training and generalization of complicated neural networks. And
 - Shake-Shake regularization can be seen as an extension of this concept where gradient noise is replaced by a form of gradient augmentation.

This is a paper in 2017 ICLR Workshop with over 10 citations. And the long version in 2017 arXiv has got over 100 citations.

**Comment by NorbertZheng** [comment]
## Shake-Shake Regularization
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/7c9f7b1b-d632-49bd-aee3-d49194fd6abc)
Left: Forward training pass. Center: Backward training pass. Right: At test time.

 - In particular, 3-branch [ResNet](https://towardsdatascience.com/review-resnet-winner-of-ilsvrc-2015-image-classification-localization-detection-e39402bfa5d8) is studied in this paper as in the above figure, with the equation as below:
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/c2a4886f-dc47-487b-89b6-aa9ce2b51dd5)

With Shake-Shake Regularization, $\alpha$ is added:
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/bfd6d32a-43b4-4b57-8cac-4f7c18677c81)

$\alpha$ is set to $0.5$ during test time, just like Dropout.

**Comment by NorbertZheng** [comment]
## Experimental Results
### CIFAR-10
26 2×32d [ResNet](https://towardsdatascience.com/review-resnet-winner-of-ilsvrc-2015-image-classification-localization-detection-e39402bfa5d8) (i.e. the network has a depth of 26, 2 residual branches and the first residual block has a width of 32) is used.
 - Forward & Backward:
   - Shake: All scaling coefficients are overwritten with new random numbers before the pass.
   - Even: All scaling coefficients are set to 0.5 before the pass.
   - Keep: For the backward pass, keep the scaling coefficients used during the forward pass.
 - Level:
    - Batch: For each residual block $i$, the same scaling coefficient is applied for all the images in the mini-batch.
    - Image: For each residual block $i$, a different scaling coefficient is applied for each image in the mini-batch.

![image](https://github.com/NorbertZheng/read-papers/assets/37235023/08887b9f-1ce4-48e7-9b13-c112ba08fe69)
Error Rates of CIFAR-10.

 - **Using Shake at forward pass has better performance.**

And Shake-Shake-Image (S-S-I) obtains the best result for 26 2×64d [ResNet](https://towardsdatascience.com/review-resnet-winner-of-ilsvrc-2015-image-classification-localization-detection-e39402bfa5d8) and 26 2×96d [ResNet](https://towardsdatascience.com/review-resnet-winner-of-ilsvrc-2015-image-classification-localization-detection-e39402bfa5d8).

**Comment by NorbertZheng** [comment]
### CIFAR-100
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/733baf5e-349f-49a8-8170-6c799b8a401c)
Error Rates of CIFAR-100.

Using Shake at forward pass again improves the performance.

Particularly, Shake-Even-Image (S-E-I) is the best.

**Comment by NorbertZheng** [comment]
### Comparison with State-of-the-art Approaches
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/f17f7520-d903-4312-9bad-d04ea09d68af)
Test error (%) and Model Size on CIFAR.

On CIFAR-10, S-S-I outperforms [WRN](https://towardsdatascience.com/review-wrns-wide-residual-networks-image-classification-d3feb3fb2004), [ResNeXt](https://towardsdatascience.com/review-resnext-1st-runner-up-of-ilsvrc-2016-image-classification-15d7f17b42ac) and [DenseNet](https://towardsdatascience.com/review-densenet-image-classification-b6631a8ef803).

On CIFAR-100, S-E-I outperforms [WRN](https://towardsdatascience.com/review-wrns-wide-residual-networks-image-classification-d3feb3fb2004), [ResNeXt](https://towardsdatascience.com/review-resnext-1st-runner-up-of-ilsvrc-2016-image-classification-15d7f17b42ac) and [DenseNet](https://towardsdatascience.com/review-densenet-image-classification-b6631a8ef803) as well.

**Comment by NorbertZheng** [comment]
## Further Evaluation
### Correlation Between Residual Branches
 - To calculate the correlation, first forward the mini-batch, through the residual branch 1 and store the output tensor in $y_{i}(1)$. Similar for residual branch 2 and store it in $y_{i}(2)$.
 - Then flatten $y_{i}(1)$ and $y_{i}(2)$ as $flat_{i}(1)$ and $flat_{i}(2)$ respectively. And calculate the covariance between each corresponding item in 2 vectors.
 - Calculate the variances of $flat_{i}(1)$ and $flat_{i}(2)$.
 - **Repeat until all images in the test set. Use the resulting covariance and variances to calculate the correlation.**

![image](https://github.com/NorbertZheng/read-papers/assets/37235023/c1d98994-8533-480a-a53e-fd294f08b6da)
Correlation results on E-E-B and S-S-I models.

 - First of all, the correlation between the output tensors of the 2 residual branches seems to be reduced by the regularization. This would support the assumption that **the regularization forces the branches to learn something different**.

![image](https://github.com/NorbertZheng/read-papers/assets/37235023/521410ce-3e1d-4f00-b383-9c20d00dd58d)
Layer-wise correlation between the first 3 layers of each residual block.

The summation at the end of the residual blocks forces an alignment of the layers on the left and right residual branches.

The correlation is reduced by the regularization.

**Comment by NorbertZheng** [comment]
### Regularization Strength
![image](https://github.com/NorbertZheng/read-papers/assets/37235023/e1644830-b049-41fe-a014-509f3eb78c8d)
Update Rules for $\beta$.

![image](https://github.com/NorbertZheng/read-papers/assets/37235023/b3e0051a-986f-46d2-9121-cd0ea1c93d39)
Left: Training curves (dark) and test curves (light) of models M1 to M5. Right: Illustration of the different methods in the above Table.

The further away $\beta$ is from $\alpha$, the stronger the regularization effect.

**Comment by NorbertZheng** [comment]
### Removing Skip Connection / Batch Normalization
 - Architecture A is 26 2×32d but without skip connection.
 - Architecture B is the same as A but with only 1 convolutional layer per branch and twice the number of blocks.
 - Architecture C is the same as A but without batch normalization.

![image](https://github.com/NorbertZheng/read-papers/assets/37235023/c8e187f0-2e11-427a-940c-2aaeb9b536c1)
Error Rates of CIFAR-10.

 - The results of architecture A clearly show that shake-shake regularization can work even without a skip connection.
 - The results of architecture B show that regularization no longer works.
 - Architecture C makes the model difficult to converge, makes the model a lot more sensitive. It is also really easy to make the model diverge.

**Comment by NorbertZheng** [comment]
With the simple yet novel idea and of course the positive results, it is published in 2017 ICLR Workshop which is very encouraging.

**Comment by NorbertZheng** [comment]
## References
 - [2017 arXiv] [Shake-Shake] [Shake-Shake Regularization](https://arxiv.org/abs/1705.07485).
 - [2017 ICLR Workshop] [Shake-Shake] [Shake-Shake Regularization of 3-Branch Residual Networks](https://pdfs.semanticscholar.org/22aa/426aeffb77339646cc03da8e94de22396efc.pdf).
